Create tokens from configuration files (#39164)
dotnet/docs: docs/azure/sdk/authentication/create-token-credentials-from-configuration.md

---
title: Create token credentials from configuration
description: This article describes how to create token credentials from configuration files
ms.topic: how-to
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 1/16/2024
---

# Create Azure credential types using configuration files

> The `Microsoft.Extensions.Azure` library supports creating different <xref:Azure.Core.TokenCredential?displayProperty=fullName> types from values defined in _appsettings.json_ and other configuration files. This article describes the support for different `TokenCredential` types and how to configure the required values for each type.

Reviewer: Consider linking to the NuGet package, we've had feedback that whenever docs mention a NuGet package - we should link to it.

## Support for Azure credentials through configuration

The `Microsoft.Extensions.Azure` library can automatically provide Azure service clients with a `TokenCredential` class by searching _appsettings.json_ or other configuration files for credential values using the default `IConfiguration` service for .NET. This approach allows developers to explicitly set credential values across different environments through configuration rather than through app code directly.

The following credential types are supported via configuration:

> * <xref:Azure.Identity.ManagedIdentityCredential?displayProperty=fullName>

Reviewer 2: I recommend alphabetizing this list and linking to the subsections in this doc instead of to the API ref pages

* <xref:Azure.Identity.WorkloadIdentityCredential?displayProperty=fullName>
* <xref:Azure.Identity.ClientSecretCredential?displayProperty=fullName>
* <xref:Azure.Identity.ClientCertificateCredential?displayProperty=fullName>
* <xref:Azure.Identity.DefaultAzureCredential?displayProperty=fullName>

The configuration file values are only used when the service client doesn't explicitly set an authentication mechanism. For example, the following code uses `IConfiguration` to search for values in _appsettings.json_ at run time because the <xref:Azure.Storage.Blobs.BlobServiceClient?displayProperty=fullName> does not specify credentials during instantiation:

```csharp
// No TokenCredential or access key provided - configuration files will be searched

> var blobServiceClient = new BlobServiceClient("<storage-account-name>");

Reviewer 3: This is not correct.  Configuration awareness is not directly built-into the Azure SDK clients.  The string-based constructor only accepts a full connection string.  

The configuration-driven credential is only used by the Azure Extensions when creating clients registered for DI.   It essentially calls the `builder.UseCredential` method to replace the default credential used for global auth or the local `builder.WithCredential` method to override the global credential for a specific client instance.

_([src](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/AzureClientFactoryBuilder.cs#L35))_

```

In comparison, the following code provides `DefaultAzureCredential` directly and therefore does *not* initiate a search for configuration values in _appsettings.json_:

```csharp
// Configuration files will not be searched - DefaultAzureCredential is already provided
var blobServiceClient = new BlobServiceClient("<storage-account-name>", new DefaultAzureCredential());
```

> [!NOTE]
> The examples in this article use `BlobServiceClient`, but the concepts apply to other Azure service clients as well, such as `CosmosClient` or `SecretClient`.

### Create a `ManagedIdentityCredential` type

Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.ManagedIdentityCredential?displayProperty=fullName>:

```json
{
  "credential": "managedidentity",

>   "resourceId":  "<managedIdentityResourceId>"

Reviewer 2: I envision two subheadings in this section:
1. User-assigned managed identity
1. System-assigned managed identity

We'll need to test this, but I think those JSON snippets would look as follows.

For user-assigned support, choose one of the following approaches:

1. Authenticate a client ID:

```json
{
  "credential": "managedidentity",
  "clientId": "<clientId>"
}
```

2. Authenticate a resource ID (the resource ID takes the form `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}`):

```json
{
  "credential": "managedidentity",
  "resourceId": "<managedIdentityResourceId>"
}
```

For system-assigned support:

```json
{
  "credential": "managedidentitycredential"
}
```

See https://github.com/Azure/azure-sdk-for-net/blob/08ac84b398ca9d6f87a70c0d5c8ec8ea9acec191/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientFactory.cs#L108-L121 for context.

---
title: Create token credentials from configuration

> description: This article describes how to create token credentials from configuration files

Reviewer 2: ```suggestion
description: This article describes how to create Microsoft Entra token credentials from configuration files.
```

ms.topic: how-to
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 1/16/2024
---

> # Create Azure credential types using configuration files

Reviewer 2: ```suggestion
# Create Microsoft Entra credential types using configuration files
```

The `Microsoft.Extensions.Azure` library supports creating different <xref:Azure.Core.TokenCredential?displayProperty=fullName> types from values defined in _appsettings.json_ and other configuration files. This article describes the support for different `TokenCredential` types and how to configure the required values for each type.

## Support for Azure credentials through configuration

> The [`Microsoft.Extensions.Azure`](https://www.nuget.org/packages/Microsoft.Extensions.Azure) library can automatically provide Azure service clients with a `TokenCredential` class by searching _appsettings.json_ or other configuration files for credential values using the default `IConfiguration` service for .NET. This approach allows developers to explicitly set credential values across different environments through configuration rather than through app code directly.

Reviewer 4: Should we link to https://learn.microsoft.com/dotnet/core/extensions/configuration also?
```suggestion
The [`Microsoft.Extensions.Azure`](https://www.nuget.org/packages/Microsoft.Extensions.Azure) library can automatically provide Azure service clients with a `TokenCredential` class by searching _appsettings.json_ or other configuration files for credential values using the `IConfiguration` abstraction for .NET. This approach allows developers to explicitly set credential values across different environments through configuration rather than through app code directly.
```

The following credential types are supported via configuration:

> * [ClientCertificateCredential](#create-a-clientcertificatecredential-type)

Reviewer 5: additionallyAllowedTenantsList applies to several of these. I think all except for WorkloadIdentityCredential and ManagedIdentityCredential.

Author: Added a section for this.

* [ClientSecretCredential](#create-a-clientsecretcredential-type)
* [DefaultAzureCredential](#create-a-defaultazurecredential-type)
* [ManagedIdentityCredential](#create-a-managedidentitycredential-type)
* [WorkloadIdentityCredential](#create-a-workloadidentitycredential-type)

> ## Configuring Azure credentials

Reviewer 2: ```suggestion
## Configure Azure credentials
```

> Azure service clients registered with the `AddAzureClients` method are automatically configured with an instance of `DefaultAzureCredential`. You can override the global `DefaultAzureCredential` using  credential values from configuration files when registering a client:

Reviewer 2: ```suggestion
Azure service clients registered with the `AddAzureClients` method are automatically configured with an instance of `DefaultAzureCredential`. You can override the global `DefaultAzureCredential` using credential values from configuration files when registering a client:
```

> Azure service clients registered with the `AddAzureClients` method are automatically configured with an instance of `DefaultAzureCredential`. You can override the global `DefaultAzureCredential` using credential values from configuration files when registering a client:

Reviewer 4: ```suggestion
Azure service clients registered with the `AddAzureClients` method are automatically configured with an instance of `DefaultAzureCredential` if no explicit credential is supplied via the `WithCredential` extension method. You can also override the global `DefaultAzureCredential` using credential values from configuration files when registering a client:
```

```csharp
builder.Services.AddAzureClients(clientBuilder =>
{
    // Register BlobServiceClient using credentials from appsettings.json
    clientBuilder.AddBlobServiceClient(builder.Configuration.GetSection("Storage"));

    // Register ServiceBusClient using the fallback DefaultAzureCredential credentials
    clientBuilder.AddServiceBusClientWithNamespace(
        "<your_namespace>.servicebus.windows.net");
});
```

The associated _appsettings.json_ file:

```json
"Storage": {
    "serviceUri": "<service_uri>",
    "credential": "managedidentity",
    "clientid":  "<clientId>"
}
```

### Create a `ManagedIdentityCredential` type

You can create both user-assigned and system-assigned managed identities using configuration values. Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.ManagedIdentityCredential?displayProperty=fullName>.

#### User-assigned identities

> 1. Authenticate using a client ID:

Reviewer 4: ```suggestion
1. Specify a user-assigned managed identity via a client ID:
```

    ```json
    {
        "credential": "managedidentity",
        "clientId":  "<clientId>"
    }
    ```

> 1. Authenticate using a resource ID:

Reviewer 4: ```suggestion
1. Specify a user-assigned managed identity via a resource ID:
```

    ```json
    {
        "credential": "managedidentity",
        "resourceId":  "<managedIdentityResourceId>"
    }
    ```

    The resource ID takes the form `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}`.

> #### System-assigned identities

Reviewer 2: I don't think system-assigned managed identities are currently supported via this configuration approach. @christothes Can you please take a look?

Reviewer 4: As long as not specifying a clientId in the config results in a null value, this should work. @alexwolfmsft can you validate with your testing?

> 1. Authenticate using a client ID:

Reviewer 4: ```suggestion
1. Configure managed identity with a system-assigned identity:
```

>     ```json

Reviewer 5: It seems like we are missing the credential heading like we have in all of the other examples, e.g. https://github.com/dotnet/docs/pull/39164/files#diff-0a55f75251b3928717aa7ca16536dd78c01b389f5409a74da159a5efd279d39aR88

    {
        "credential": "managedidentity",

>         "clientid":  "<clientId>"

Reviewer 4: This should be omitted to select a system-assigned MI. Related to the comment above, we should validate that this works (I think it would).

Author: I'm not sure I follow - that would mean the only value would be "credential": "managedIdetity". Based on this code below it seems like you have to provide a resourceId or clientId to create a managedIdentityCredential:

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientFactory.cs#L108-L120

Author: I'm also not sure how to test this scenario since the fallback is DefaultAzureCredential, so even if specifically a system assigned or user assigned through configuration doesn't get picked up properly, it would still work using the fallback DefaultAzureCredential?

Reviewer 4: To clarify - set  `"credential": "managedidentity"` and omit `clientid`

Author: Do you have any suggestions on how to test this? If I deploy an app with a registered service client to Azure I'm not sure how to verify whether its creating a system assigned identity using the config file or if its just falling back to DefaultAzureCredential since all of this happens internally.

Reviewer 4: Nothing about the config will create a system assigned identity - it's just whether or not the config specifying `"credential": "managedidentity"` without a clientId will succeed. If it does, this should mean that we created a DAC with no options, which defaults to system-assigned identity.

Author: Sorry if I'm not understanding this right - if you specify "credential": "managedidentity" without a clientId I believe that would end up at the line linked below and pass in a null value for clientId, which would create a ManagedIdentityCredential, not DefaultAzureCredential - is this what you'd expect?

Reviewer 4: >  is this what you'd expect?

Yes. Do you expect something different?

Author: No, just making sure I understood your previous message - thanks.

    }
    ```

### Create a `WorkloadIdentityCredential` type

Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.WorkloadIdentityCredential?displayProperty=fullName>:

```json
{
    "credential": "workloadidentity",
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",
    "tokenFilePath": "<tokenFilePath>"
}
```

### Create a `ClientSecretCredential` type

Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.ClientSecretCredential?displayProperty=fullName>:

```json
{
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",

>     "tokenFilePath": "<clientSecret>"

Reviewer 5: ```suggestion
    "clientSecret": "<clientSecret>"
```

}
```

### Create a `ClientCertificateCredential` type

Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.ClientCertificateCredential?displayProperty=fullName>:

```json
{
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",

>     "clientCertificate": "<clientCertificate>"

Reviewer 5: Should we also document the optional store name and location - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientFactory.cs#L95-L96

Reviewer 2: Yes, we should

---
title: Create token credentials from configuration
description: This article describes how to create Microsoft Entra token credentials from configuration files.
ms.topic: how-to
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 1/16/2024
---

# Create Microsoft Entra credential types using configuration files

> The `Microsoft.Extensions.Azure` library supports creating different <xref:Azure.Core.TokenCredential?displayProperty=fullName> types from values defined in _appsettings.json_ and other configuration files. This article describes the support for different `TokenCredential` types and how to configure the required values for each type.

Reviewer 2: ```suggestion
The `Microsoft.Extensions.Azure` library supports creating different <xref:Azure.Core.TokenCredential?displayProperty=fullName> types from key-value pairs defined in _appsettings.json_ and other configuration files. The credential types correspond to a subset of the [credential classes](/dotnet/api/overview/azure/identity-readme?view=azure-dotnet#credential-classes) in the Azure Identity client library. This article describes the support for different `TokenCredential` types and how to configure the required values for each type.
```

> The `Microsoft.Extensions.Azure` library supports creating different <xref:Azure.Core.TokenCredential?displayProperty=fullName> types from key-value pairs defined in _appsettings.json_ and other configuration files. The credential types correspond to a subset of the [credential classes](/dotnet/api/overview/azure/identity-readme?view=azure-dotnet#credential-classes) in the Azure Identity client library. This article describes the support for different `TokenCredential` types and how to configure the required values for each type.

Reviewer 2: ```suggestion
The `Microsoft.Extensions.Azure` library supports creating different <xref:Azure.Core.TokenCredential?displayProperty=fullName> types from key-value pairs defined in _appsettings.json_ and other configuration files. The credential types correspond to a subset of the [credential classes](/dotnet/api/overview/azure/identity-readme?view=azure-dotnet#credential-classes) in the Azure Identity client library. This article describes the support for different `TokenCredential` types and how to configure the required key-value pairs for each type.
```

## Support for Azure credentials through configuration

> The [`Microsoft.Extensions.Azure`](https://www.nuget.org/packages/Microsoft.Extensions.Azure) library can automatically provide Azure service clients with a `TokenCredential` class by searching _appsettings.json_ or other configuration files for credential values using the `IConfiguration` abstraction for .NET. This approach allows developers to explicitly set credential values across different environments through configuration rather than through app code directly.

Reviewer 2: ```suggestion
The [`Microsoft.Extensions.Azure`](https://www.nuget.org/packages/Microsoft.Extensions.Azure) library can automatically provide Azure service clients with a `TokenCredential` class by searching _appsettings.json_, _appsettings.{environment}.json_, or other configuration files for credential values using .NET's `IConfiguration` abstraction. This approach allows developers to explicitly set credential values across different environments through configuration rather than through app code directly.
```

```csharp
builder.Services.AddAzureClients(clientBuilder =>
{
    // Register BlobServiceClient using credentials from appsettings.json
    clientBuilder.AddBlobServiceClient(builder.Configuration.GetSection("Storage"));

    // Register ServiceBusClient using the fallback DefaultAzureCredential credentials
    clientBuilder.AddServiceBusClientWithNamespace(
        "<your_namespace>.servicebus.windows.net");
});
```

The associated _appsettings.json_ file:

```json
"Storage": {
    "serviceUri": "<service_uri>",
    "credential": "managedidentity",
    "clientId":  "<clientId>"
}
```

The following credential types also support the `AdditionallyAllowedTenants` property, which specifies additional Microsoft Entra tenants beyond the default tenant for which the credential may acquire tokens:

* [ClientCertificateCredential](#create-a-clientcertificatecredential-type)
* [ClientSecretCredential](#create-a-clientsecretcredential-type)
* [DefaultAzureCredential](#create-a-defaultazurecredential-type)

>  Add the wildcard value "*" to allow the credential to acquire tokens for any Microsoft Entra tenant the logged in account can access. If no tenant IDs are specified, this option will have no effect on that authentication method, and the credential will acquire tokens for any requested tenant when using that method.

Reviewer 2: ```suggestion
Add the wildcard value "*" to allow the credential to acquire tokens for any Microsoft Entra tenant the logged in account can access. If no tenant IDs are specified, this option will have no effect on that authentication method, and the credential will acquire tokens for any requested tenant when using that method.
```

```json
{
    "additionallyAllowedTenants":  "<tenant-ids-separated-by-semicolon>"
}
```

### Create a `ManagedIdentityCredential` type

> You can create both user-assigned and system-assigned managed identities using configuration values. Add the following configuration values to your _appsettings.json_ file to create an instance of <xref:Azure.Identity.ManagedIdentityCredential?displayProperty=fullName>.

Reviewer 2: ```suggestion
You can create both user-assigned and system-assigned managed identities using configuration values. Add the following key-value pairs to your _appsettings.json_ file to create an instance of <xref:Azure.Identity.ManagedIdentityCredential?displayProperty=fullName>.
```

#### User-assigned managed identities

> 1. Specify a user-assigned managed identity via a client ID:

Reviewer 2: I think this should be a bulleted list instead. We should also clarify that you choose one approach or the other.

```json
{
    "credential": "managedidentity",
    "clientId":  "<clientId>"
}
```

Specify a user-assigned managed identity via a resource ID:

```json
{
    "credential": "managedidentity",
    "managedIdentityResourceId":  "<managedIdentityResourceId>"
}
```

The resource ID takes the form `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}`.

#### System-assigned managed identities

```json
{
    "credential": "managedidentity"
}
```

### Create a `WorkloadIdentityCredential` type

> Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.WorkloadIdentityCredential?displayProperty=fullName>:

Reviewer 2: ```suggestion
Add the following key-value pairs to your _appsettings.json_ file to create an <xref:Azure.Identity.WorkloadIdentityCredential?displayProperty=fullName>:
```

```json
{
    "credential": "workloadidentity",
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",
    "tokenFilePath": "<tokenFilePath>"
}
```

### Create a `ClientSecretCredential` type

> Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.ClientSecretCredential?displayProperty=fullName>:

Reviewer 2: ```suggestion
Add the following key-value pairs to your _appsettings.json_ file to create an <xref:Azure.Identity.ClientSecretCredential?displayProperty=fullName>:
```

```json
{
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",
    "clientSecret": "<clientSecret>"
}
```

### Create a `ClientCertificateCredential` type

> Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.ClientCertificateCredential?displayProperty=fullName>:

Reviewer 2: ```suggestion
Add the following key-value pairs to your _appsettings.json_ file to create an <xref:Azure.Identity.ClientCertificateCredential?displayProperty=fullName>:
```

```json
{
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",
    "clientCertificate": "<clientCertificate>",
    "clientCertificateStoreLocation": "<clientCertificateStoreLocation>",
    "additionallyAllowedTenants": "<tenant-ids-separated-by-semicolon>"
}
```

> [!NOTE]

> > The `clientCertificateStoreLocation` and `additionallyAllowedTenants` values are optional. If no `clientCertificateStoreLocation` is specified, the default `CurrentUser` is used from the <xref:System.Fabric.X509Credentials.StoreLocation?displayProperty=nameWithType> enum.

Reviewer 2: What happens if the `clientCertificateStoreLocation` and `additionallyAllowedTenants` keys are present, but their values are empty strings? Is that treated the same as the absence of both key-value pairs? It would be good to say here.

### Create a `DefaultAzureCredential` type

> Add the following configuration values to your _appsettings.json_ file to create an <xref:Azure.Identity.DefaultAzureCredential?displayProperty=fullName>:

Reviewer 2: ```suggestion
Add the following key-value pairs to your _appsettings.json_ file to create an <xref:Azure.Identity.DefaultAzureCredential?displayProperty=fullName>:
```

## Support for Azure credentials through configuration

The [`Microsoft.Extensions.Azure`](https://www.nuget.org/packages/Microsoft.Extensions.Azure) library can automatically provide Azure service clients with a `TokenCredential` class by searching _appsettings.json_ or other configuration files for credential values using the `IConfiguration` abstraction for .NET. This approach allows developers to explicitly set credential values across different environments through configuration rather than through app code directly.

The following credential types are supported via configuration:

* [ClientCertificateCredential](#create-a-clientcertificatecredential-type)
* [ClientSecretCredential](#create-a-clientsecretcredential-type)
* [DefaultAzureCredential](#create-a-defaultazurecredential-type)
* [ManagedIdentityCredential](#create-a-managedidentitycredential-type)
* [WorkloadIdentityCredential](#create-a-workloadidentitycredential-type)

## Configure Azure credentials

> Azure service clients registered with the `AddAzureClients` method are automatically configured with an instance of `DefaultAzureCredential` if no explicit credential is supplied via the `WithCredential` extension method. You can also override the global `DefaultAzureCredential` using credential values from configuration files when registering a client to create a specific credential type:

Reviewer 2: Link to the API ref pages for `AddAzureClients` and `WithCredential`

```csharp
builder.Services.AddAzureClients(clientBuilder =>
{
    // Register BlobServiceClient using credentials from appsettings.json
    clientBuilder.AddBlobServiceClient(builder.Configuration.GetSection("Storage"));

    // Register ServiceBusClient using the fallback DefaultAzureCredential credentials
    clientBuilder.AddServiceBusClientWithNamespace(
        "<your_namespace>.servicebus.windows.net");
});
```

The associated _appsettings.json_ file:

```json
"Storage": {
    "serviceUri": "<service_uri>",
    "credential": "managedidentity",

>     "clientid":  "<clientId>"

Reviewer 2: ```suggestion
    "clientId":  "<clientId>"
```

}
```

> The following credential types also support the `AdditionallyAllowedTenants` property, which specifies additional tenants beyond the default tenant for which the credential may acquire tokens:

Reviewer 2: ```suggestion
The following credential types also support the `AdditionallyAllowedTenants` property, which specifies additional Microsoft Entra tenants beyond the default tenant for which the credential may acquire tokens:
```

* [ClientCertificateCredential](#create-a-clientcertificatecredential-type)
* [ClientSecretCredential](#create-a-clientsecretcredential-type)
* [DefaultAzureCredential](#create-a-defaultazurecredential-type)

 Add the wildcard value "*" to allow the credential to acquire tokens for any tenant the logged in account can access. If no value tenant Ids are specified, this option will have no effect on that authentication method, and the credential will acquire tokens for any requested tenant when using that method.

```json
{
    "additionallyAllowedTenants":  "<tenant-ids-separated-by-semicolon>"
}
```

### Create a `ManagedIdentityCredential` type

You can create both user-assigned and system-assigned managed identities using configuration values. Add the following configuration values to your _appsettings.json_ file to create an instance of <xref:Azure.Identity.ManagedIdentityCredential?displayProperty=fullName>.

> #### User-assigned identities

Reviewer 2: ```suggestion
#### User-assigned managed identities
```

1. Specify a user-assigned managed identity via a client ID:

    ```json
    {
        "credential": "managedidentity",
        "clientId":  "<clientId>"
    }
    ```

1. Specify a user-assigned managed identity via a resource ID:

    ```json
    {
        "credential": "managedidentity",
        "managedIdentityResourceId":  "<managedIdentityResourceId>"
    }
    ```

    The resource ID takes the form `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}`.

> #### System-assigned identities

Reviewer 2: ```suggestion
#### System-assigned managed identities
```

{
    "tenantId":  "<tenantId>",
    "clientId":  "<clientId>",
    "clientCertificate": "<clientCertificate>",
    "clientCertificateStoreLocation": "<clientCertificateStoreLocation>",
    "additionallyAllowedTenants": "<tenant-ids-separated-by-semicolon>"
}
```

> [!NOTE]

> > The `clientCertificateStoreLocation` and `additionallyAllowedTenants` values are optional.

Reviewer 4: It may be worth adding that if no `clientCertificateStoreLocation` is specified, the default `CurrentUser` is used (reference of cert store location values is [here](https://learn.microsoft.com/dotnet/api/system.security.cryptography.x509certificates.storelocation))
